### DISCUSSION TITLE
Nullable settings in ctor of `BlockGridItem`

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#14550] [Features and ideas]
I have this following where a Header can be used on its own, but also works as composition on existing blocks, e.g. a few carousel blocks.

I can render the tagline and a heading properties directly, or include a partial view using `IBlockHeader` model in each block inclusive the Header block, but I found the following works to render the block directly.

However should SettingsUdi and Settings arguments be nullable as these are not required?

```
@if (content is IBlockHeader header)
{
    var blockHeader = new BlockHeader(header, _iPublishedValueFallback);

    var blockItem = new Umbraco.Cms.Core.Models.Blocks.BlockGridItem<BlockHeader>(Model.ContentUdi, blockHeader, null, null);

    @await Html.PartialAsync("BlockGrid/Components/" + BlockHeader.ModelTypeAlias, blockItem)

    @*if (!string.IsNullOrEmpty(header.Tagline))
    {
        <div class="tagline subheading">@header.Tagline</div>
    }

    if (!string.IsNullOrEmpty(header.Heading))
    {
        <h2>@header.Heading</h2>
    }*@
}
```

![image](https://github.com/umbraco/Umbraco-CMS/assets/2919859/1643ba6f-9cf7-48f8-bbac-6dd92dcaf3a4)

**Comment by abjerner** [comment]
Constructor looks like as shown below, so might have been an oversight:

```csharp
public BlockGridItem(Udi contentUdi, IPublishedElement content, Udi settingsUdi, IPublishedElement settings)
{
    ContentUdi = contentUdi ?? throw new ArgumentNullException(nameof(contentUdi));
    Content = content ?? throw new ArgumentNullException(nameof(content));
    SettingsUdi = settingsUdi;
    Settings = settings;
}
```

So I agree that it makes sense to make them nullable. Although ideally I suppose `settingsUdi` must not be null if `settings` is not null (and vice versa).

It could probably also make sense to add a constructor overload that only takes the two first parameters:

```csharp
public BlockGridItem(Udi contentUdi, IPublishedElement content)
{
    ContentUdi = contentUdi ?? throw new ArgumentNullException(nameof(contentUdi));
    Content = content ?? throw new ArgumentNullException(nameof(content));
}
```

**Comment by bjarnef** [comment]
Yeah, I guess the settings should be nullable, maybe @Zeegaan can answer that?

I also wonder it we should have an extension like this?

```
@Html.GetBlockGridItemHtml(Umbraco.Cms.Core.Models.Blocks.BlockGridItem<T> item, string partialViewName)
@Html.GetBlockGridItemHtmlAsync(Umbraco.Cms.Core.Models.Blocks.BlockGridItem<T> item, string partialViewName)
```

but wouldn't benefits much it the code above.

Not sure it we can make rendering a single block item from a composition simpler.
I guess for now we just have the handle this part ourselves:

```
var blockHeader = new BlockHeader(header, _iPublishedValueFallback);
var blockItem = new Umbraco.Cms.Core.Models.Blocks.BlockGridItem<BlockHeader>(Model.ContentUdi, blockHeader, null, null);
```

**Comment by Zeegaan** [comment]
I haven't looked into this, but at a quick glance, my best bet would be:  yes they should be nullable 👍
